lib/ansiblelint/errors.py

```diff
@@ -27,3 +34,19 @@ def __repr__(self):
 
         return formatstr.format(_id, self.message,
                                 self.filename, self.linenumber, self.line)
+
+    @property
+    def _hash_key(self):
+        return self.filename, self.linenumber, self.message, getattr(self.rule, 'id', 0)
+
+    def __lt__(self, other):
+        """Enable sorting of MatchError instances."""
+        return self._hash_key < other._hash_key
+
+    def __hash__(self):
+        """Perform hash of matches."""
+        return hash(self._hash_key)
+
+    def __eq__(self, other):
+        """Identify duplicate matches."""
+        return self.__hash__() == other.__hash__()
```

Comment:
Hash equality is not enough. We should check if the type is compatible too. And if not, we should give a chance to the other object to check it.
```suggestion
        if not isinstance(other, self.__class__):
            raise NotImplementedError
        return self.__hash__() == other.__hash__()
```